eth/gasprice/gasprice.go

```diff
@@ -64,82 +71,101 @@ func NewOracle(backend ethapi.Backend, params Config) *Oracle {
+		sampleNumber: sampleNumber,
+		percentile:   percent,
 	}
 }
 
-// SuggestPrice returns the recommended gas price.
-func (gpo *Oracle) SuggestPrice(ctx context.Context) (*big.Int, error) {
+// NewFullOracle returns a gasprice oracle for full node which has
+// avaiblable recent blocks in local db. FullOracle has higher
+// recommendation accuracy.
+func NewFullOracle(backend ethapi.Backend, params Config) *Oracle {
+	return newOracle(backend, fullSampleNumber, params)
+}
+
+// NewLightOracle returns a gasprice oracle for light client which doesn't
+// has recent block locally. LightOracle is much cheaper than FullOracle
+// however the corresponding recommendation accuracy is lower.
+func NewLightOracle(backend ethapi.Backend, params Config) *Oracle {
+	return newOracle(backend, lightSampleNumber, params)
+}
+
+// getLatest returns a recommended gas price which is suggested last time
+// but still suitable now.
+func (gpo *Oracle) getLatest(headHash common.Hash) *big.Int {
 	gpo.cacheLock.RLock()
 	lastHead := gpo.lastHead
 	lastPrice := gpo.lastPrice
 	gpo.cacheLock.RUnlock()
+	if headHash == lastHead {
+		return lastPrice
+	}
+	return nil
+}
 
+// SuggesstPrice returns a gasprice so that newly created transaction can
+// has very high chance to be included in the following blocks.
+func (gpo *Oracle) SuggestPrice(ctx context.Context) (*big.Int, error) {
 	head, _ := gpo.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
 	headHash := head.Hash()
-	if headHash == lastHead {
-		return lastPrice, nil
+	// Firstly check whether there is available gasprice for recommendation.
+	if price := gpo.getLatest(headHash); price != nil {
+		return price, nil
 	}
-
 	gpo.fetchLock.Lock()
 	defer gpo.fetchLock.Unlock()
-
-	// try checking the cache again, maybe the last fetch fetched what we need
-	gpo.cacheLock.RLock()
-	lastHead = gpo.lastHead
-	lastPrice = gpo.lastPrice
-	gpo.cacheLock.RUnlock()
-	if headHash == lastHead {
-		return lastPrice, nil
+	// Try checking the cache again, maybe the last fetch fetched what we need
+	if price := gpo.getLatest(headHash); price != nil {
+		return price, nil
 	}
-
-	blockNum := head.Number.Uint64()
-	ch := make(chan getBlockPricesResult, gpo.checkBlocks)
-	sent := 0
-	exp := 0
-	var blockPrices []*big.Int
-	for sent < gpo.checkBlocks && blockNum > 0 {
-		go gpo.getBlockPrices(ctx, types.MakeSigner(gpo.backend.ChainConfig(), big.NewInt(int64(blockNum))), blockNum, ch)
+	var (
+		sent, exp int
+		number    = head.Number.Uint64()
+		ch        = make(chan getBlockPricesResult, gpo.checkBlocks)
+		txPrices  []*big.Int
+	)
+	for sent < gpo.checkBlocks && number > 0 {
+		go gpo.getBlockPrices(ctx, types.MakeSigner(gpo.backend.ChainConfig(), big.NewInt(int64(number))), number, gpo.sampleNumber, ch)
 		sent++
 		exp++
-		blockNum--
+		number--
 	}
-	maxEmpty := gpo.maxEmpty
+	maxInvalid := gpo.maxInvalid
 	for exp > 0 {
 		res := <-ch
 		if res.err != nil {
-			return lastPrice, res.err
+			return gpo.lastPrice, res.err
```

Comment:
Another reason to not use `getLatest`: now you are accessing the `gpo` struct directly while the cache lock is not held anymore (potential data race).

Reply:
Yes you are right